### PR TITLE
#34685 Collapse search results one level at a time

### LINKED ISSUE #34685: Search: collapse but keep roots expanded
One pattern I am seeing myself doing a lot is invoking the "Collapse" action on the search viewlet and now that I have multi root results, I see no results until I expand a root:

image

Could we maybe collapse all but leave the roots expanded? Or, upon first collapse call we do that and upon the second we collapse also the roots?

### PR DESCRIPTION
**Pull request by RMacfarlane**
Fixes [#34685](https://github.com/Microsoft/vscode/issues/34685)

Changes the behavior of the "Collapse All" button to collapse results one level at a time instead of immediately to root.

### CODE REVIEW

File: src/vs/base/parts/tree/browser/tree.ts [hunk 1]
```diff
@@ -96,10 +96,10 @@ export interface ITree extends Events.IEventEmitter {
 
 	/**
 	 * Collapses several elements.
-	 * Provide no arguments and it will recursively collapse all elements in the tree
+	 * Collapses all elements at the greatest tree depth that has expanded elements.
 	 * The returned promise returns a boolean for whether the elements were collapsed or not.
 	 */
-	collapseAll(elements?: any[], recursive?: boolean): WinJS.Promise;
+	collapseDeepestExpandedLevel(): WinJS.Promise;
```
Comment (roblourens) [inline src/vs/base/parts/tree/browser/tree.ts]: Let's leave `collapseAll` in place, having `collapseDeepestExpandedLevel` as a new method

File: src/vs/base/parts/tree/browser/treeModel.ts [hunk 2]
```diff
@@ -882,18 +893,29 @@ export class TreeModel extends Events.EventEmitter {
...
-		for (var i = 0, len = elements.length; i < len; i++) {
-			promises.push(this.collapse(elements[i], recursive));
+	public collapseDeepestExpandedLevel(): WinJS.Promise {
+		var levelToCollapse = this.findDeepestExpandedLevel(this.input, 0);
+
+		var items = [this.input];
+		for (var i = 0; i < levelToCollapse; i++) {
+			items = items
+				.map(node => node.getChildren())
+				.reduce((prev, current) => prev.concat(current), []);
```
Comment (roblourens) [inline src/vs/base/parts/tree/browser/treeModel.ts]: We have a helper for this, `arrays.flatten`

File: src/vs/workbench/browser/viewlet.ts [hunk 3]
```diff
@@ -259,12 +259,12 @@ export class ToggleViewletAction extends Action {
...
 
 	constructor(viewer: ITree, enabled: boolean, clazz: string) {
-		super('workbench.action.collapse', nls.localize('collapse', "Collapse All"), clazz, enabled, (context: any) => {
+		super('workbench.action.collapse', nls.localize('collapse', "Collapse"), clazz, enabled, (context: any) => {
 			if (viewer.getHighlight()) {
 				return TPromise.as(null); // Global action disabled if user is in edit mode from another action
 			}
 
-			viewer.collapseAll();
+			viewer.collapseDeepestExpandedLevel();
```
Comment (roblourens) [inline src/vs/workbench/browser/viewlet.ts]: I think we have two actions that do the same thing. Let's replace one with an action that calls `collapseDeepestExpandedLevel`, and leave the other. And let's be explicit in which trees were are changing the behavior of. We should change it for search results in this PR and maybe the file explorer. If we want other trees to have this behavior, we can change that later.

**Review by roblourens — CHANGES_REQUESTED** [review]

File: src/vs/base/parts/tree/browser/treeDefaults.ts [hunk 4]
```diff
@@ -427,7 +427,7 @@ export class CollapseAllAction extends Action {
 			return TPromise.as(null); // Global action disabled if user is in edit mode from another action
 		}
 
-		this.viewer.collapseAll();
+		this.viewer.collapseDeepestExpandedLevel();
```
Comment (roblourens) [inline src/vs/base/parts/tree/browser/treeDefaults.ts]: This action shouldn't be called CollapseAllAction anymore

File: src/vs/workbench/browser/viewlet.ts (hunk 3 above)
Comment (RMacfarlane) [reply src/vs/workbench/browser/viewlet.ts]: I've scoped it to just search results for now

**Review by roblourens — APPROVED** [review]

**Comment by roblourens** [comment]
Thanks!

**Pull request merged**